gnome-libs-1.4.1.4/man/gnome-mime.5.in
.\"
.\" GNOME mime types manual page
.\" (C) 1999 Miguel de Icaza (miguel@gnu.org).
.\" 
.\" This manual page is covered by the terms of the GNU General
.\" Public License.  
.\"
.TH GNOME 1 "GNOME 1.0" 
.SH NAME
gnome-mime \- configuration files
.SH DESCRIPTION
.PP
The GNOME system uses MIME types to classify content.  Each MIME type
on the GNOME system has a number of attributes attached to it.  Some
of these attributes have a special meaning to the GNOME system.
.SH MIME type of files
There are two ways of classifying a file in the GNOME system: by
matching their extension or a regular expression with their name or by
its content.  GNOME applications use one of those two methods
depending on speed contraints and the specific setup.  The MIME types
available on the system can be configured at runtime by putting
special files in either the GNOME MIME directory
.I (@datadir@/mime-info)
or the user 
.I ~/.gnome/mime-info 
directory.
.PP
.SH MIME type definition files.
.PP
The routines that classify a file by its name, use the contents of all
of the files with the extension 
.I .mime
from the
.I @datadir@/mime-info
directory and the
.I ~/.gnome/mime-info
to build the database for
filename matching.  The latter is supported to enable users to provide
their mime types to extend the system defaults.
.PP
Application that wish to install their own MIME types only need
to install a file in this directory.
.PP
The file 
.I @datadir@/mime-info/gnome.mime 
is
special, as it contains the defaults for gnome, and is read first.  In
addition, the file 
.I ~/.gnome/mime-info/user.mime
is read last.  This will guarantee that there is a way to set system
defaults, and there is a way for the user to override them.  There is
currently no way to tell anything about the order of the other files
in those directories, nor is there anyway to override system defaults
yet.
.PP
The format is the following:
.PP
.nf
mime-type-name
	ext[,priority]: ext1 ext2 ext3
	ext[,priority]: ext4
	regex[,priority]: regex1
	regex[,priority]: regex2
.fi
.PP
where "mime-type-name" is a valid MIME type.  For example
"text/plain". 
.PP
For example, for a vCalendar application, this file would be
installed:
.PP
.nf
------ calendar.mime  -------
application/v-calendar:
	ext: vcf
-----------------------------
.fi
.PP
.SH MIME key information
To add keys to a MIME type, it is necessary to install a file with the
extension 
.I .keys
in the
.I @datadir@/mime-info
directory or in the 
.I ~/.gnome/mime-info
directory.  The former is for system-provided mime-information and the
latter is to enable the user to extend the actions as provided by the
system.  
.PP
The file 
.I @datadir@/mime-info/gnome.keys
is
special, as it contains the defaults for gnome, and is read first.  In
addition, the file 
.I ~/.gnome/mime-info/user.keys
is
read last.  This will guarantee that there is a way to set system
defaults, and there is a way for the user to override them.  There is
currently no way to tell anything about the order of the other files in
those directories, nor is there anyway to override system defaults yet.
.PP
The 
.I .keys
files have the following format:
.PP
.nf
mime-type-match:
	[\[LANG\]]key=value
.fi
.PP
Above, the 
.I key
is the key that is being
defined and 
.I value
is the value we bind to
it.  The optional 
.I [LANG]
represents a language in which this definition is valid.  If this part
is specified, then the definition will only be valid if LANG matches
the setting of the environment variable LANG.  The LANG setting is
used to provide keys which can be displayed to the user in a localized
way.
.PP
This is an example to bind the key 
.I open
to all of
the mime-types matching 
.I image/*
and the
.I icon-filename
key is bound to the
.I /opt/gimp/share/xcf.png
value:
.PP
.nf
image/*:
	open=gimp %f

image/x-xcf:
        icon-filename=/opt/gimp/share/xcf.png
.fi
.PP
This will make the GIMP the handler for the open action.  Files of
type xcf would use the filename pointed in the 
.I icon-filename
key.
.PP
.I %f gets
interpolated with the file name or the list of file names that
matched this mime-type.
.PP
As you can see from the example above, a
.I .keys
file does not need to provide all of the values, it can just provide
or override some of the actions.
.PP
User defined bindings in 
.I .keys file will take
precedence over system installed files.
.PP
.SH Special key used by the GNOME system
The following keys are currently used in the GNOME
desktop:
.PP
.I open
.IP
Open the file with this command.
.PP
.I icon-filename
.IP
The filename with the icon that should be used to represent files of this type.
.PP
.I view
.IP
Command to view the file contents.
.PP
.I ascii-view
.IP
A command that should be used to do an ascii-rendering of the file.
Used as a fallback by the filemanager if a
.I view action
does not exist.
.PP
.I fm-open
.IP
file-manager open.  If
present, the file manager will use this action instead of the value in
open to perform this action (the filemanager for example will open
archive files as if they were directories by using the VFS).
.PP
.I fm-view
.IP
file-manager view.  If present, invoking the view opertion on the file
manager will use the value defined here instead of the value in
"view".
.PP
.I fm-ascii-view
.IP
Fallback operation for the file manager as well.
.PP
Those keys are also queried on the metadata (except in the cases where
the lookup would be too expensive).
.PP
.SH AUTHOR
This manual page has been written by Miguel de Icaza (miguel@gnu.org)



eel-1.0.2/NEWS
Version 1.0.2:

* Greatly speed up preference handling by storing visibility and
  default values internally rather than using GConf (Ramiro)
* Use a 64K buffer instead of a 4K one for faster image loading (Darin)
* Speed up some bad bottlenecks in scalable font handling (Alex)

* Fix bugs that would treat dark backgrounds as light (Darin)
* Fix storage leaks (Anders, Darin)
* Make EelList work better as a replacement for GtkCList (Maciej)
* Create the trash directory when needed during drag (Jonathan)
* Fix problems in multibyte locales (James Su, Alex)
* Set up directories so you can install eel for gnome 1 and eel for
  gnome 2 on the same machine (Ramiro)
* Translate messages properly by using a different gettext for
  library vs. application messages (Darin)
* Don't segfault if a NULL MIME type is returned by gnome-vfs (Darin)

* Use a special drop type instead of a special file name for resetting
  the background -- this means that the reset feature won't work until
  you get a new Nautilus (Andy)

* Make drop feedback nicer for EelCTree by bolding text as well as
  darkening the icon (Maciej)

* More flexibility in the preferences class (Ramiro)
* Add more EelStringList functions (Ramiro)
* Add feature where EelImage can render at a lower opacity when
  the widget is insensitive if requested (Ramiro)
* Add more ArtPoint and ArtDRect functions (Ramiro)

Version 1.0.1:

* Speed up drawing of solid backgrounds. (Alex Larsson)
* Speed up handling of default fonts. (Alex Larsson)
* Switched from Alt to Control for list view key bindings to
  match changes in Nautilus and Sawfish. (John Harper)
* Fixed x/y problem that was causing a bug in Nautilus list view
  clicking. (Darin)
* Improve handling of URIs and escaped characters. (Darin)
* Fix some return_if_fails that were incorrect. (Darin, Ramiro)
* Make font manager code work even when there are tabs in the
  font control files. (Eungkyu Song, Darin)
* Make the font manager code work when fonts are symbolic links.
  (Frederic Devernay, Ramiro).
* Added more calls and clean up some implementation details,
  including more self check machinery. (Ramiro)


gnome-libs-1.4.1.4/libgnome/gnome-mime.c
/*
 * Copyright 1997 Paolo Molaro
 * Copyright 1998 Miguel de Icaza.
 *
 */

#include <config.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <dirent.h>
#include <regex.h>
#include "libgnomeP.h"
#include "gnome-mime.h"
#include <string.h>
#include <ctype.h>

static gboolean module_inited = FALSE;

static GHashTable *mime_extensions [2] = { NULL, NULL };
static GList      *mime_regexs     [2] = { NULL, NULL };

typedef struct {
	regex_t regex;
	char    *mime_type;
} RegexMimePair;

typedef struct {
	char *dirname;
	struct stat s;
	unsigned int valid : 1;
	unsigned int system_dir : 1;
} mime_dir_source_t;

/* These ones are used to automatically reload mime-types on demand */
static mime_dir_source_t gnome_mime_dir, user_mime_dir;
static time_t last_checked;

#ifdef G_THREADS_ENABLED

/* We lock this mutex whenever we modify global state in this module.  */
G_LOCK_DEFINE_STATIC (mime_mutex);

#endif /* G_LOCK_DEFINE_STATIC */



static char *
get_priority (char *def, int *priority)
{
	*priority = 0;

	if (*def == ','){
		def++;
		if (*def == '1'){
			*priority = 0;
			def++;
		} else if (*def == '2'){
			*priority = 1;
			def++;
		}
	}

	while (*def && *def == ':')
		def++;

	return def;
}

static void
add_to_key (char *mime_type, char *def)
{
	int priority = 1;
	char *s, *p, *ext;

	if (strncmp (def, "ext", 3) == 0){
		char *tokp;

		def += 3;
		def = get_priority (def, &priority);
		s = p = g_strdup (def);

		while ((ext = strtok_r (s, " \t\n\r,", &tokp)) != NULL){
			gpointer res = g_hash_table_lookup (mime_extensions [priority], ext);
			if (res == NULL)
			g_hash_table_insert (mime_extensions [priority], g_strdup (ext), g_strdup (mime_type));
#if 0
			else if (strcmp (mime_type, res))
				printf ("%s = %s while inserting %s\n", ext, (char *)res, mime_type);
#endif
			s = NULL;
		}
		g_free (p);
	}

	if (strncmp (def, "regex", 5) == 0){
		RegexMimePair *mp;
		def += 5;
		def = get_priority (def, &priority);

		while (*def && isspace (*def))
			def++;

		if (!*def)
			return;

		mp = g_new (RegexMimePair, 1);
		if (regcomp (&mp->regex, def, REG_EXTENDED | REG_NOSUB)){
			g_free (mp);
			return;
		}
		mp->mime_type = g_strdup (mime_type);

		mime_regexs [priority] = g_list_prepend (mime_regexs [priority], mp);
	}
}

static void
mime_fill_from_file (const char *filename)
{
	FILE *f;
	char buf [1024];
	char *current_key;

	g_assert (filename != NULL);

	f = fopen (filename, "r");

	if (!f)
		return;

	current_key = NULL;
	while (fgets (buf, sizeof (buf), f)){
		char *p;

		if (buf [0] == '#')
			continue;

		/* Trim trailing spaces */
		for (p = buf + strlen (buf) - 1; p >= buf; p--){
			if (isspace (*p) || *p == '\n')
				*p = 0;
			else
				break;
		}

		if (!buf [0])
			continue;

		if (buf [0] == '\t' || buf [0] == ' '){
			if (current_key){
				char *p = buf;

				while (*p && isspace (*p))
					p++;

				if (*p == 0)
					continue;

				add_to_key (current_key, p);
			}
		} else {
			if (current_key)
				g_free (current_key);

			current_key = g_strdup (buf);
			if (current_key [strlen (current_key)-1] == ':')
				current_key [strlen (current_key)-1] = 0;
		}
	}

	if (current_key)
		g_free (current_key);

	fclose (f);
}

static void
mime_load (mime_dir_source_t *source)
{
	DIR *dir;
	struct dirent *dent;
	const int extlen = sizeof (".mime") - 1;
	char *filename;

	g_return_if_fail (source != NULL);
	g_return_if_fail (source->dirname != NULL);

	if (stat (source->dirname, &source->s) != -1)
		source->valid = TRUE;
	else
		source->valid = FALSE;

	dir = opendir (source->dirname);
	if (!dir){
		source->valid = FALSE;
		return;
	}

	if (source->system_dir){
		filename = g_concat_dir_and_file (source->dirname, "gnome.mime");
		mime_fill_from_file (filename);
		g_free (filename);
	}

	while ((dent = readdir (dir)) != NULL){

		int len = strlen (dent->d_name);

		if (len <= extlen)
			continue;
		if (strcmp (dent->d_name + len - extlen, ".mime"))
			continue;

		if (source->system_dir && !strcmp (dent->d_name, "gnome.mime"))
			continue;
		if (!source->system_dir && !strcmp (dent->d_name, "user.mime"))
			continue;

		filename = g_concat_dir_and_file (source->dirname, dent->d_name);

		mime_fill_from_file (filename);
		g_free (filename);
	}
	closedir (dir);

	if (!source->system_dir) {
		filename = g_concat_dir_and_file (source->dirname, "user.mime");
		mime_fill_from_file (filename);
		g_free (filename);
	}
}

static gboolean
mime_hash_func (gpointer key, gpointer value, gpointer user_data)
{
	g_free (key);
	g_free (value);

	return TRUE;
}

static void
maybe_reload (void)
{
	time_t now = time (NULL);
	gboolean need_reload = FALSE;
	struct stat s;
	int i;

	if (last_checked + 5 >= now)
		return;

	if (stat (gnome_mime_dir.dirname, &s) != -1)
		if (s.st_mtime != gnome_mime_dir.s.st_mtime)
			need_reload = TRUE;

	if (stat (user_mime_dir.dirname, &s) != -1)
		if (s.st_mtime != user_mime_dir.s.st_mtime)
			need_reload = TRUE;

	last_checked = now;

	if (!need_reload)
		return;

	for (i = 0; i < 2; i++){
		GList *l;

		g_hash_table_foreach_remove (mime_extensions [i], mime_hash_func, NULL);

		for (l = mime_regexs [i]; l; l = l->next){
			RegexMimePair *mp = l->data;

			g_free (mp->mime_type);
			regfree (&mp->regex);
			g_free (mp);
		}
		g_list_free (mime_regexs [i]);
		mime_regexs [i] = NULL;
	}

	mime_load (&gnome_mime_dir);
	mime_load (&user_mime_dir);

	last_checked = time (NULL);
}

static void
mime_init (void)
{
	mime_extensions [0] = g_hash_table_new (g_str_hash, g_str_equal);
	mime_extensions [1] = g_hash_table_new (g_str_hash, g_str_equal);

	gnome_mime_dir.dirname = gnome_unconditional_datadir_file ("mime-info");
	gnome_mime_dir.system_dir = TRUE;

	user_mime_dir.dirname  = g_concat_dir_and_file (gnome_util_user_home (), ".gnome/mime-info");
	user_mime_dir.system_dir = FALSE;
	mime_load (&gnome_mime_dir);
	mime_load (&user_mime_dir);

	last_checked = time (NULL);
	module_inited = TRUE;
}

/**
 * gnome_mime_type_or_default:
 * @filename: A filename (the file does not necesarily exist).
 * @defaultv: A default value to be returned if no match is found
 *
 * This routine tries to determine the mime-type of the filename
 * only by looking at the filename from the GNOME database of mime-types.
 *
 * Returns the mime-type of the @filename.  If no value could not be
 * determined, it will return @defaultv.
 */
const char *
gnome_mime_type_or_default (const gchar *filename, const gchar *defaultv)
{
	const gchar *ext;
	int priority;
	const gchar *result = defaultv;

	G_LOCK (mime_mutex);

	if (!filename)
		goto done;
	ext = strrchr (filename, '.');
	if (ext)
		++ext;

	if (!module_inited)
		mime_init ();

	maybe_reload ();

	for (priority = 1; priority >= 0; priority--){
		GList *l;
		char *res;

		if (ext){
			gchar *upext;

			res = g_hash_table_lookup (mime_extensions [priority], ext);
			if (res) {
				result = res;
				goto done;
			}
			/* Search for the UPPER case extension */
			upext = g_strdup (ext);
			g_strup (upext);
			res = g_hash_table_lookup (mime_extensions [priority],
						   upext);
			if (res) {
				result = res;
				g_free (upext);
				goto done;
			}

			/* Final check for lower case */
			g_strdown (upext);
			res = g_hash_table_lookup (mime_extensions [priority],
						   upext);
			g_free (upext);
			if (res) {
				result = res;
				goto done;
			}
		}

		for (l = mime_regexs [priority]; l; l = l->next){
			RegexMimePair *mp = l->data;

			if (regexec (&mp->regex, filename, 0, 0, 0) == 0) {
				result = mp->mime_type;
				goto done;
			}
		}
	}

 done:
	G_UNLOCK (mime_mutex);
	return result;
}

/**
 * gnome_mime_type:
 * @filename: A filename (the file does not necesarily exist).
 *
 * Determined the mime type for @filename.
 *
 * Returns the mime-type for this filename.
 */
const char *
gnome_mime_type (const gchar * filename)
{
	return gnome_mime_type_or_default (filename, "text/plain");
}

/**
 * gnome_mime_type_or_default_of_file:
 * @existing_filename: A filename that points to an existing filename.
 * @defaultv: A default value to be returned if no match is found
 *
 * This routine tries to determine the mime-type of the filename
 * by trying to guess the content of the file.  If this fails, it will
 * return the mime-type based only on the filename.
 *
 * Returns the mime-type of the @existing_filename.  If no value could not be
 * determined, it will return @defaultv.
 */
const char *
gnome_mime_type_or_default_of_file (const char *existing_filename,
				    const gchar *defaultv)
{
	char *mime_type;

	mime_type = (char *)gnome_mime_type_from_magic (existing_filename);
	if (mime_type)
		return mime_type;

	return gnome_mime_type_or_default (existing_filename, defaultv);
}

/**
 * gnome_mime_type_of_file:
 * @existing_filename: A filename pointing to an existing file.
 *
 * Determined the mime type for @existing_filename.  It will try
 * to figure this out by looking at the contents of the file, if this
 * fails it will use the filename to figure out a name.
 *
 * Returns the mime-type for this filename.
 */
const char *
gnome_mime_type_of_file (const char *existing_filename)
{
	return gnome_mime_type_or_default_of_file (existing_filename, "text/plain");
}


/**
 * gnome_uri_list_extract_uris:
 * @uri_list: an uri-list in the standard format.
 *
 * Returns a GList containing strings allocated with g_malloc
 * that have been splitted from @uri-list.
 */
GList*
gnome_uri_list_extract_uris (const gchar* uri_list)
{
	const gchar *p, *q;
	gchar *retval;
	GList *result = NULL;

	g_return_val_if_fail (uri_list != NULL, NULL);

	p = uri_list;

	/* We don't actually try to validate the URI according to RFC
	 * 2396, or even check for allowed characters - we just ignore
	 * comments and trim whitespace off the ends.  We also
	 * allow LF delimination as well as the specified CRLF.
	 */
	while (p) {
		if (*p != '#') {
			while (isspace(*p))
				p++;

			q = p;
			while (*q && (*q != '\n') && (*q != '\r'))
				q++;

			if (q > p) {
			        q--;
				while (q > p && isspace(*q))
					q--;

				retval = g_malloc (q - p + 2);
				strncpy (retval, p, q - p + 1);
				retval[q - p + 1] = '\0';

				result = g_list_prepend (result, retval);
			}
		}
		p = strchr (p, '\n');
		if (p)
			p++;
	}

	return g_list_reverse (result);
}

/**
 * gnome_uri_list_extract_filenames:
 * @uri_list: an uri-list in the standard format
 *
 * Returns a GList containing strings allocated with g_malloc
 * that contain the filenames in the uri-list.
 *
 * Note that unlike gnome_uri_list_extract_uris() function, this
 * will discard any non-file uri from the result value.
 */
GList*
gnome_uri_list_extract_filenames (const gchar* uri_list)
{
	GList *tmp_list, *node, *result;

	g_return_val_if_fail (uri_list != NULL, NULL);

	result = gnome_uri_list_extract_uris (uri_list);

	tmp_list = result;
	while (tmp_list) {
		gchar *s = tmp_list->data;

		node = tmp_list;
		tmp_list = tmp_list->next;

		if (!strncmp (s, "file:", 5)) {
			node->data = g_strdup (s+5);
		} else {
			result = g_list_remove_link(result, node);
			g_list_free_1 (node);
		}
		g_free (s);
	}
	return result;
}

/**
 * gnome_uri_list_free_strings:
 * @list: A GList returned by gnome_uri_list_extract_uris() or gnome_uri_list_extract_filenames()
 *
 * Releases all of the resources allocated by @list.
 */
void
gnome_uri_list_free_strings      (GList *list)
{
	g_list_foreach (list, (GFunc)g_free, NULL);
	g_list_free (list);
}
